ext/sysvshm/sysvshm.c

```diff
@@ -154,40 +186,40 @@ PHP_FUNCTION(shm_attach)
 		chunk_ptr->free = shm_size-chunk_ptr->end;
 	}
 
+	shm_list_ptr = Z_SYSVSHM_P(return_value);
+
 	shm_list_ptr->key = shm_key;
 	shm_list_ptr->id = shm_id;
 	shm_list_ptr->ptr = chunk_ptr;
-
-	RETURN_RES(zend_register_resource(shm_list_ptr, php_sysvshm.le_shm));
 }
 /* }}} */
 
-/* {{{ proto bool shm_detach(resource shm_identifier)
+/* {{{ proto bool shm_detach(Sysvshm shm_identifier)
    Disconnects from shared memory segment */
 PHP_FUNCTION(shm_detach)
 {
 	zval *shm_id;
-	sysvshm_shm *shm_list_ptr;
 
-	if (SUCCESS != zend_parse_parameters(ZEND_NUM_ARGS(), "r", &shm_id)) {
+	if (SUCCESS != zend_parse_parameters(ZEND_NUM_ARGS(), "O", &shm_id, sysvshm_ce)) {
 		RETURN_THROWS();
 	}
-	SHM_FETCH_RESOURCE(shm_list_ptr, shm_id);
-	RETURN_BOOL(SUCCESS == zend_list_close(Z_RES_P(shm_id)));
+
+	RETURN_TRUE;
```

Comment:
Hm, this looks problematic. I think it would be better to do the shmdt() here and set the pointer to null.

Reply:
This is what you mean or I'm just groping in the dark?
```
shmdt((void *) shm_list_ptr->ptr);
shm_list_ptr->ptr = NULL;
``

Reply:
Yes, this is what I had in mind. Though this also means that need to check for NULL ptr in other functions.

Reply:
I'm wondering why this change is necessary? Originally, the function just marked the memory segment destroyable, but did not destroy it until `shm_detach()` was called. Now, `shm_remove()` will both mark destroyable and destroy. Is this acceptable? Isn't it better to do the same in `shm_detach()` (and when the objects is GCd)?

Reply:
I think this code just ended up in the wrong place? What I had in mind is to have the shmdt call in shm_detach and leave shm_remove as it is now.

Reply:
Yes, you are right, somehow I made the changes to the wrong function.